flink-ml-iteration/src/main/java/org/apache/flink/iteration/datacache/nonkeyed/DataCacheWriter.java

```diff
@@ -19,127 +19,162 @@
+            currentWriter.finish().ifPresent(finishedSegments::add);
+            currentWriter = new FileSegmentWriter<>(serializer, pathGenerator.get());
+            Preconditions.checkState(currentWriter.addRecord(record));
+        }
     }
 
+    /** Finishes adding records and closes resources occupied for adding records. */
     public List<Segment> finish() throws IOException {
-        finishCurrentSegment(false);
-        return finishSegments;
-    }
+        if (currentWriter == null) {
+            return finishedSegments;
+        }
 
-    public FileSystem getFileSystem() {
-        return fileSystem;
+        currentWriter.finish().ifPresent(finishedSegments::add);
+        currentWriter = null;
+        return finishedSegments;
     }
 
-    public List<Segment> getFinishSegments() {
-        return finishSegments;
+    /**
+     * Flushes all added records to segments and returns a list of segments containing all cached
+     * records.
+     */
+    public List<Segment> getSegments() throws IOException {
+        finishCurrentSegmentIfAny();
+        return finishedSegments;
     }
 
-    private void finishCurrentSegment(boolean newSegment) throws IOException {
-        if (currentSegment != null) {
-            currentSegment.finish().ifPresent(finishSegments::add);
-            currentSegment = null;
+    private void finishCurrentSegmentIfAny() throws IOException {
+        if (currentWriter == null) {
+            return;
         }
 
-        if (newSegment) {
-            currentSegment = new SegmentWriter(pathGenerator.get());
-        }
+        currentWriter.finish().ifPresent(finishedSegments::add);
+        currentWriter = createSegmentWriter();
     }
 
-    private class SegmentWriter {
-
-        private final Path path;
-
-        private final FSDataOutputStream outputStream;
-
-        private final DataOutputView outputView;
-
-        private int currentSegmentCount;
-
-        public SegmentWriter(Path path) throws IOException {
-            this.path = path;
-            this.outputStream = fileSystem.create(path, FileSystem.WriteMode.NO_OVERWRITE);
-            this.outputView = new DataOutputViewStreamWrapper(outputStream);
+    /** Cleans up all previously added records. */
+    public void cleanup() throws IOException {
+        finishCurrentSegmentIfAny();
+        for (Segment segment : finishedSegments) {
+            if (segment.getFsSize() > 0) {
+                fileSystem.delete(segment.getPath(), false);
+            }
+            if (segment.getCache() != null) {
+                segmentPool.returnAll(segment.getCache());
+            }
         }
+        finishedSegments.clear();
+    }
 
-        public void addRecord(T record) throws IOException {
-            serializer.serialize(record, outputView);
-            currentSegmentCount += 1;
-        }
+    /** Persists the segments in this writer to disk. */
+    public void persistSegmentsToDisk() throws IOException {
```

Comment:
nits: given that we use `FileSegmentWriter`, would it be more consistent to rename this method as `writeSegmentsToFiles()`?